File: src/src/main/java/com/q42/imagepreview/ImagePreview.java
package com.q42.imagepreview;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.renderscript.Allocation;
import android.renderscript.Element;
import android.renderscript.RenderScript;
import android.renderscript.ScriptIntrinsicBlur;
import android.support.annotation.NonNull;
import android.util.Base64;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Arrays;

public final class ImagePreview {

    private static final byte[] SOF0PATTERN = {(byte) 0xFF, (byte) 0xC0};
    private static byte[] HEADER = null;

    /**
     * Private constructor
     * There is no need for this class to be extended or instantiated from the outside
     */
    private ImagePreview() {}

    /**
     * Returns an Android Drawable object gained from inserting the header into this
     * static class using the function setHeader(String) and inserting the body
     * into this function.
     *
     * @param body       {String} the Base64 encoded body of the image
     * @param context    {Context} the context from which to create a drawable
     * @param blurRadius {Float} Optional blur amount
     * @return {Drawable} The Android Drawable containing the image
     * @throws IOException
     */
    @SuppressLint("NewApi")
    public static Drawable drawableFromString(@NonNull String body, Context context, Float blurRadius) throws IOException, IllegalArgumentException {

        final Bitmap preview = bitmapFromString(body);

        if(preview == null) {
            return null;
        }

        if (blurRadius != null && Build.VERSION.SDK_INT > Build.VERSION_CODES.JELLY_BEAN_MR1) {

            final Bitmap blurred = Bitmap.createBitmap(preview);

            RenderScript rs = RenderScript.create(context);
            ScriptIntrinsicBlur theIntrinsic = ScriptIntrinsicBlur.create(rs, Element.U8_4(rs));
            Allocation tmpOut = Allocation.createFromBitmap(rs, blurred);

            theIntrinsic.setRadius(blurRadius);
            theIntrinsic.setInput(Allocation.createFromBitmap(rs, preview));
            theIntrinsic.forEach(tmpOut);

            tmpOut.copyTo(blurred);

            return new BitmapDrawable(context.getResources(), blurred);
        }

        return new BitmapDrawable(context.getResources(), preview);
    }

    public static Drawable drawableFromString(@NonNull String body, Context context) throws IOException, IllegalArgumentException {
        return drawableFromString(body, context, null);
    }

    /**
     * Returns an Android Bitmap object gained from inserting the header into this
     * static class using the function setHeader(String) and inserting the body
     * into this function.
     *
     * @param body {String} the Base64 encoded body of the image
     * @return {Bitmap} The Android Bitmap containing the image
     * @throws IOException
     */
    public static Bitmap bitmapFromString(@NonNull String body) throws IOException, IllegalArgumentException {

        final byte[] byteImage = byteImageFromBody(body);
        if (byteImage == null) {
            return null;
        }

        return BitmapFactory.decodeByteArray(byteImage, 0, byteImage.length);
    }

    /**
     * Insert the Base64 encoded body of the image and get back the base64 encoded
     * image including the header.
     *
     * @param inputBody {String} The body of the image base64 encoded
     * @return {byte[]} The image base64 encoded with the header included or null if something failed
     * @throws IOException
     */
    public static byte[] byteImageFromBody(@NonNull String inputBody) throws IOException, IllegalArgumentException {

        if (HEADER == null || inputBody == null) {
            return null;
        }

        final int indexC0 = ImagePreview.indexOf(HEADER, SOF0PATTERN);
        if (indexC0 == -1) {
            return null;
        }

        final int headerSizeIndexStart = indexC0 + 5;
        final byte[] body = Base64.decode(inputBody, Base64.NO_WRAP);

        if(body == null || body.length < 5) {
            return null;
        }

        final ByteArrayOutputStream finalArray = new ByteArrayOutputStream();

        finalArray.write(Arrays.copyOfRange(getHeader(), 0, headerSizeIndexStart));
        finalArray.write(Arrays.copyOfRange(body, 1, 5));
        finalArray.write(Arrays.copyOfRange(getHeader(), headerSizeIndexStart, getHeader().length));
        finalArray.write(Arrays.copyOfRange(body, 5, body.length));

        return finalArray.toByteArray();
    }

    /**
     * Get the base64 decoded header string previously set
     * May return null if no header is set yet.
     *
     * @return {byte[]} the header as a byte array
     */
    public static byte[] getHeader() {
        return HEADER;
    }

    /**
     * Set the header which is then decoded into a
     * Base64 byte array used to parse the image eventually
     *
     * @param header {String} the still encoded Base64 string image header
     */
    public static void setHeader(@NonNull String header) {
        HEADER = Base64.decode(header, Base64.NO_WRAP);
    }

    /**
     * Function found on the interwebs: http://stackoverflow.com/a/21341168
     * Used to find the index of a byte array in a byte array.
     *
     * @param outerArray   {byte[]} The array to search in
     * @param smallerArray {byte[]} The array to search for
     * @return The position or -1 if not found
     */
    public static int indexOf(byte[] outerArray, byte[] smallerArray) {

        if (outerArray == null || smallerArray == null || outerArray.length == 0 ||
                smallerArray.length == 0 || outerArray.length < smallerArray.length) {
            // Guard against input that might throw  nasty runtime exceptions!
            return -1;
        }

        for (int i = 0; i < outerArray.length - smallerArray.length + 1; ++i) {
            boolean found = true;
            for (int j = 0; j < smallerArray.length; ++j) {
                if (outerArray[i + j] != smallerArray[j]) {
                    found = false;
                    break;
                }
            }
            if (found) {
                return i;
            }
        }

        return -1;
    }
}

File: src/src/main/java/com/q42/imagepreview/ImagePreview.kt
@file:JvmName("ImagePreview")

package com.q42.imagepreview

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.drawable.BitmapDrawable
import android.os.Build
import android.renderscript.Allocation
import android.renderscript.Element
import android.renderscript.RenderScript
import android.renderscript.ScriptIntrinsicBlur
import android.util.Base64
import java.io.ByteArrayOutputStream

/**
 * Created by thijs on 28-06-16.
 */

private val headers = mapOf(1 to "/9j/4AAQSkZJRgABAQEAYABgAAD/2wBDAAoHBwkHBgoJCAkLCwoMDxkQDw4ODx4WFxIZJCAmJSMgIyIoLTkwKCo2KyIjMkQyNjs9QEBAJjBGS0U+Sjk/QD3/2wBDAQsLCw8NDx0QEB09KSMpPT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT3/wAARCAMBIgACEQEDEQH/xAAfAAABBQEBAQEBAQAAAAAAAAAAAQIDBAUGBwgJCgv/xAC1EAACAQMDAgQDBQUEBAAAAX0BAgMABBEFEiExQQYTUWEHInEUMoGRoQgjQrHBFVLR8CQzYnKCCQoWFxgZGiUmJygpKjQ1Njc4OTpDREVGR0hJSlNUVVZXWFlaY2RlZmdoaWpzdHV2d3h5eoOEhYaHiImKkpOUlZaXmJmaoqOkpaanqKmqsrO0tba3uLm6wsPExcbHyMnK0tPU1dbX2Nna4eLj5OXm5+jp6vHy8/T19vf4+fr/xAAfAQADAQEBAQEBAQEBAAAAAAAAAQIDBAUGBwgJCgv/xAC1EQACAQIEBAMEBwUEBAABAncAAQIDEQQFITEGEkFRB2FxEyIygQgUQpGhscEJIzNS8BVictEKFiQ04SXxFxgZGiYnKCkqNTY3ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZnaGlqc3R1dnd4eXqCg4SFhoeIiYqSk5SVlpeYmZqio6Slpqeoqaqys7S1tre4ubrCw8TFxsfIycrS09TV1tfY2dri4+Tl5ufo6ery8/T19vf4+fo=".decodeBase64())

@JvmOverloads fun Context.previewDrawable(body: String, blurRadius: Int = 4) = previewBitmap(body, blurRadius)?.let { BitmapDrawable(this.resources, it) }

@JvmOverloads fun Context.previewBitmap(body: String, blurRadius: Int = 4): Bitmap? {
    val bitmap = bitmapFromString(body) ?: return null

    if (blurRadius == 0) {
        return bitmap
    }

    return try {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.JELLY_BEAN_MR1) {
            renderScriptBlur(bitmap, blurRadius, this)
        } else {
            stackBlur(bitmap, blurRadius)
        }
    } catch(e: Exception) {
        null
    }
}

@SuppressLint("NewApi")
private fun renderScriptBlur(bitmap: Bitmap, blurRadius: Int, context: Context): Bitmap? {
    val blurred = Bitmap.createBitmap(bitmap.width, bitmap.height, bitmap.config)

    val rs = RenderScript.create(context)
    val theIntrinsic = ScriptIntrinsicBlur.create(rs, Element.U8_4(rs))
    val tmpOut = Allocation.createFromBitmap(rs, blurred)

    theIntrinsic.setRadius(blurRadius.toFloat())
    theIntrinsic.setInput(Allocation.createFromBitmap(rs, bitmap))
    theIntrinsic.forEach(tmpOut)

    tmpOut.copyTo(blurred)

    return blurred
}

private fun bitmapFromString(bodyString: String): Bitmap? {
    val body = bodyString.decodeBase64() ?: return null

    val versionLength = 1
    val version = body[0].toInt()

    val header = headers[version] ?: return null

    val index = header.indexOf(0xFF.toByte(), 0xC0.toByte())

    if (index == -1) return null

    val headerSizeIndexStart = index + 5

    val finalArray = object : ByteArrayOutputStream(body.size + header.size) {
        val bufDirect: ByteArray get() = buf // Allow direct access to the buffer without copying
    }

    finalArray.write(header, 0, headerSizeIndexStart)
    finalArray.write(body, versionLength, 4)
    finalArray.write(header, headerSizeIndexStart, header.size - headerSizeIndexStart)
    finalArray.write(body, 4 + versionLength, body.size - 4 - versionLength)

    return try {
        BitmapFactory.decodeByteArray(finalArray.bufDirect, 0, finalArray.bufDirect.size)
    } catch(e: Exception) {
        null
    }
}

private fun String.decodeBase64() = try {
    Base64.decode(this, Base64.NO_WRAP)
} catch(e: Exception) {
    null
}

private fun ByteArray.indexOf(vararg smallerArray: Byte): Int {
    if (this.size == 0 || smallerArray.size == 0 || this.size < smallerArray.size) {
        return -1
    }

    for (i in 0..this.size - smallerArray.size) {
        var found = true
        for (j in smallerArray.indices) {
            if (this[i + j] != smallerArray[j]) {
                found = false
                break
            }
        }
        if (found) {
            return i
        }
    }

    return -1
}

private fun stackBlur(sentBitmap: Bitmap, radius: Int): Bitmap? {

    // Stack Blur v1.0 from
    // http://www.quasimondo.com/StackBlurForCanvas/StackBlurDemo.html
    //
    // Java Author: Mario Klingemann <mario at quasimondo.com>
    // http://incubator.quasimondo.com
    // created Feburary 29, 2004
    // Android port : Yahel Bouaziz <yahel at kayenko.com>
    // http://www.kayenko.com
    // ported april 5th, 2012

    // This is a compromise between Gaussian Blur and Box blur
    // It creates much better looking blurs than Box Blur, but is
    // 7x faster than my Gaussian Blur implementation.
    //
    // I called it Stack Blur because this describes best how this
    // filter works internally: it creates a kind of moving stack
    // of colors whilst scanning through the image. Thereby it
    // just has to add one new block of color to the right side
    // of the stack and remove the leftmost color. The remaining
    // colors on the topmost layer of the stack are either added on
    // or reduced by one, depending on if they are on the right or
    // on the left side of the stack.
    //
    // If you are using this algorithm in your code please add
    // the following line:
    //
    // Stack Blur Algorithm by Mario Klingemann <mario@quasimondo.com>

    val bitmap = sentBitmap.copy(sentBitmap.config, true)

    if (radius < 1) {
        return null
    }

    val w = bitmap.width
    val h = bitmap.height

    val pix = IntArray(w * h)
    bitmap.getPixels(pix, 0, w, 0, 0, w, h)

    val wm = w - 1
    val hm = h - 1
    val wh = w * h
    val div = radius + radius + 1

    val r = IntArray(wh)
    val g = IntArray(wh)
    val b = IntArray(wh)
    var rsum: Int
    var gsum: Int
    var bsum: Int
    var x: Int
    var y: Int
    var i: Int
    var p: Int
    var yp: Int
    var yi: Int
    var yw: Int
    val vmin = IntArray(Math.max(w, h))

    var divsum = div + 1 shr 1
    divsum *= divsum
    val dv = IntArray(256 * divsum)
    i = 0
    while (i < 256 * divsum) {
        dv[i] = i / divsum
        i++
    }

    yw = 0; yi = 0

    val stack = Array(div) { IntArray(3) }
    var stackpointer: Int
    var stackstart: Int
    var sir: IntArray
    var rbs: Int
    val r1 = radius + 1
    var routsum: Int
    var goutsum: Int
    var boutsum: Int
    var rinsum: Int
    var ginsum: Int
    var binsum: Int

    y = 0
    while (y < h) {
        rinsum = 0; ginsum = 0; binsum = 0; routsum = 0; goutsum = 0; boutsum = 0; rsum = 0; gsum = 0; bsum = 0

        i = -radius
        while (i <= radius) {
            p = pix[yi + Math.min(wm, Math.max(i, 0))]
            sir = stack[i + radius]
            sir[0] = p and 0xff0000 shr 16
            sir[1] = p and 0x00ff00 shr 8
            sir[2] = p and 0x0000ff
            rbs = r1 - Math.abs(i)
            rsum += sir[0] * rbs
            gsum += sir[1] * rbs
            bsum += sir[2] * rbs
            if (i > 0) {
                rinsum += sir[0]
                ginsum += sir[1]
                binsum += sir[2]
            } else {
                routsum += sir[0]
                goutsum += sir[1]
                boutsum += sir[2]
            }
            i++
        }
        stackpointer = radius

        x = 0
        while (x < w) {

            r[yi] = dv[rsum]
            g[yi] = dv[gsum]
            b[yi] = dv[bsum]

            rsum -= routsum
            gsum -= goutsum
            bsum -= boutsum

            stackstart = stackpointer - radius + div
            sir = stack[stackstart % div]

            routsum -= sir[0]
            goutsum -= sir[1]
            boutsum -= sir[2]

            if (y == 0) {
                vmin[x] = Math.min(x + radius + 1, wm)
            }
            p = pix[yw + vmin[x]]

            sir[0] = p and 0xff0000 shr 16
            sir[1] = p and 0x00ff00 shr 8
            sir[2] = p and 0x0000ff

            rinsum += sir[0]
            ginsum += sir[1]
            binsum += sir[2]

            rsum += rinsum
            gsum += ginsum
            bsum += binsum

            stackpointer = (stackpointer + 1) % div
            sir = stack[stackpointer % div]

            routsum += sir[0]
            goutsum += sir[1]
            boutsum += sir[2]

            rinsum -= sir[0]
            ginsum -= sir[1]
            binsum -= sir[2]

            yi++
            x++
        }
        yw += w
        y++
    }
    x = 0
    while (x < w) {
        rinsum = 0; ginsum = 0; binsum = 0; routsum = 0; goutsum = 0; boutsum = 0; rsum = 0; gsum = 0; bsum = 0
        yp = -radius * w
        i = -radius
        while (i <= radius) {
            yi = Math.max(0, yp) + x

            sir = stack[i + radius]

            sir[0] = r[yi]
            sir[1] = g[yi]
            sir[2] = b[yi]

            rbs = r1 - Math.abs(i)

            rsum += r[yi] * rbs
            gsum += g[yi] * rbs
            bsum += b[yi] * rbs

            if (i > 0) {
                rinsum += sir[0]
                ginsum += sir[1]
                binsum += sir[2]
            } else {
                routsum += sir[0]
                goutsum += sir[1]
                boutsum += sir[2]
            }

            if (i < hm) {
                yp += w
            }
            i++
        }
        yi = x
        stackpointer = radius
        y = 0
        while (y < h) {
            // Preserve alpha channel: ( 0xff000000 & pix[yi] )
            pix[yi] = 0xff000000.toInt() and pix[yi] or (dv[rsum] shl 16) or (dv[gsum] shl 8) or dv[bsum]

            rsum -= routsum
            gsum -= goutsum
            bsum -= boutsum

            stackstart = stackpointer - radius + div
            sir = stack[stackstart % div]

            routsum -= sir[0]
            goutsum -= sir[1]
            boutsum -= sir[2]

            if (x == 0) {
                vmin[y] = Math.min(y + r1, hm) * w
            }
            p = x + vmin[y]

            sir[0] = r[p]
            sir[1] = g[p]
            sir[2] = b[p]

            rinsum += sir[0]
            ginsum += sir[1]
            binsum += sir[2]

            rsum += rinsum
            gsum += ginsum
            bsum += binsum

            stackpointer = (stackpointer + 1) % div
            sir = stack[stackpointer]

            routsum += sir[0]
            goutsum += sir[1]
            boutsum += sir[2]

            rinsum -= sir[0]
            ginsum -= sir[1]
            binsum -= sir[2]

            yi += w
            y++
        }
        x++
    }

    bitmap.setPixels(pix, 0, w, 0, 0, w, h)

    return bitmap
}
